osu.Game/Graphics/Backgrounds/BeatmapBackground.cs

```diff
@@ -35,5 +96,209 @@ public override bool Equals(Background other)
             return other.GetType() == GetType()
                    && ((BeatmapBackground)other).Beatmap == Beatmap;
         }
+
+        public interface IColouredDimmable : IDrawable
+        {
+            Colour4 DimColour { get; set; }
+
+            float DimLevel { get; set; }
+        }
+
+        public partial class DimmableBufferedContainer : BufferedContainer, IColouredDimmable
+        {
+            private Colour4 dimColour;
+
+            private float dimLevel;
+
+            public Colour4 DimColour
+            {
+                get => dimColour;
+                set
+                {
+                    dimColour = value;
+                    Invalidate(Invalidation.DrawNode);
+                }
+            }
+
+            public float DimLevel
+            {
+                get => dimLevel;
+                set
+                {
+                    dimLevel = value;
+                    Invalidate(Invalidation.DrawNode);
+                }
+            }
+
+            public DimmableBufferedContainer(RenderBufferFormat[] formats = null, bool pixelSnapping = false, bool cachedFrameBuffer = false)
+                : base(formats, pixelSnapping, cachedFrameBuffer)
+            {
+                DimColour = Colour4.Black;
+                DimLevel = 0.0f;
+            }
+
+            [BackgroundDependencyLoader]
+            private void load(ShaderManager shaders)
+            {
+                TextureShader = shaders.Load(VertexShaderDescriptor.TEXTURE_2, "BeatmapBackground");
```

Comment:
Isn't `TextureShader` responsible for rendering the framebuffer texture of the buffered container? If so, isn't that pointless? Isn't it enough to apply the custom "BeatmapBackground" shader on the underlying beatmap background sprite? Would that cause issues with blurring? I can't imagine it will.

Reply:
Yes, but it will require the redraw of the framebuffer texture on each DimLevel change, which will also force the blur to be redrawn, and I feared that this might introduce unnecessary overhead during (for example) breaks, if player has background blur enabled. If that is not a concern - yeah, keeping the custom shader only on the sprite makes complete sense.

Reply:
You're right, it's pointless, at least in the current implementation.
Changing the dim parameters requires DrawNode invalidation, which forces the framebuffer to be redrawn, so there are no benefits. This does mean that DimLevel transitions are way more expensive now (at least on my rather old and slow hardware).

Here are the results of my benchmarks for average frametime during different parts of the beatmap.
With user blur set to 0%:
| |Gameplay|Breaks|Fade in|Fade out|
|---|---|---|---|---|
|Baseline|4.25 ms|4.29 ms|4.42 ms|4.08 ms|
|Coloured dim|4.26 ms|4.33 ms|4.92 ms|4.55 ms|
|Relative increase|0.2%|0.9%|**11.3%**|**11.5%**|

With user blur set to 100%:
| |Gameplay|Breaks|Fade in|Fade out|
|---|---|---|---|---|
|Baseline|4.20 ms|4.24 ms|4.39 ms|4.02 ms|
|Coloured dim|4.22 ms|4.30 ms|5.18 ms|4.66 ms|
|Relative increase|0.5%|1.4%|**18%**|**15.9%**|

"Baseline" is osu! 2024.1009.1
"Coloured dim" is commit [c5867b6](https://github.com/ppy/osu/pull/30391/commits/c5867b6dd9db22935e574cb2d8e7755125ece597)

<details>
<summary>Details about the benchmark</summary>

Made a map with 64 gameplay sections and 63 breaks for this benchmark, total duration is 3:09.

"Fade in" refers to frames rendered during gameplay-to-break transition, "Fade out" refers to break-to-gameplay transition.
"Gameplay" and "Breaks" columns are just for sanity-checking, as they should be same.

Total amount of recorded frames:
| |Blur|Gameplay|Breaks|Fade in|Fade out|
|---|---|---|---|---|---|
|Baseline|0%|12068|8176|11018|12069|
|Baseline|100%|12193|8253|11089|12236|
|Coloured dim|0%|12090|8105|9810|10689|
|Coloured dim|100%|12154|8137|9448|10512|

osu! settings:
- Compiled in release mode
- Single-threaded mode
- "OpenGL (Experimental)" renderer
- Windowed screen mode
- Resolution: 1564x964
- Frame limiter: Basically unlimited

System info:
- OS: Kubuntu 24.04.1 LTS x86_64
- CPU: Intel i3-6100U @ 2.300GHz
- GPU: Intel HD Graphics 520
- RAM: 19907MiB

</details>

So I guess now there are two ways to go - ditch the shader replacement on the BufferedContainer and have slower transitions, or modify the BufferedContainer somehow to avoid redrawing the framebuffer when shader parameters are changed.

Which one is more preferable?

Reply:
Fair enough, this is a bit of a pesky situation. You can control whether the framebuffer gets redrawn by overriding `GetDrawVersion()` and make that point to a custom invalidation ID that is normally incremented but not when dim properties are changed. See `Path` in osu!framework, it might actually be doing exactly what you're looking for.

Reply:
Added a test case for it, but solving it requires a change in the framework: https://github.com/ppy/osu-framework/pull/6404

With that change applied benchmarks look a lot better.

With user blur set to 0%:
|Implementation|Gameplay|Breaks|Fade in|Fade out|
|---|---|---|---|---|
|Baseline|4.15 ms|4.20 ms|4.34 ms|3.99 ms|
|Coloured dim|4.19 ms|4.23 ms|4.39 ms|4.00 ms|
|Relative increase|1.0%|0.7%|1.2%|0.3%|

With user blur set to 100%:
|Implementation|Gameplay|Breaks|Fade in|Fade out|
|---|---|---|---|---|
|Baseline|4.11 ms|4.16 ms|4.31 ms|3.93 ms|
|Coloured dim|4.05 ms|4.11 ms|4.25 ms|3.85 ms|
|Relative increase|-1.5%|-1.2%|-1.4%|-2.0%|

"Baseline" is osu! 2024.1009.1
"Coloured dim" is commit https://github.com/ppy/osu/pull/30391/commits/93c3fac26e55f90f34ff625afa0aec0e4e9f9080 with https://github.com/ppy/osu-framework/pull/6404 applied

Reply:
I don't think these benchmark numbers represent the true impact because as far as I read you're measuring over a long period of time, while when the blur re-renders it's going to be very detrimental only to the local frames (so the overall "increase" is much smaller than the actual perceived difference to the user).

Of course, as long as the numbers are going down this doesn't matter so much, so carry on I guess.

A better way of benchmarking would be to force the redraw every frame then just take a reading of the average frame time.

Reply:
I've double checked, and without https://github.com/ppy/osu-framework/pull/6404 my implementation triggered a blur redraw on every frame during transition, but didn't affect frames during gameplay or break.
This means that the stats computed in the "Fade in" and "Fade out" columns only take into account frames where redraw happened.

> A better way of benchmarking would be to force the redraw every frame then just take a reading of the average frame time.
My goal was to show that extra redraws were triggered and that it affected the performance during specific parts of the beatmap, so I couldn't use this approach